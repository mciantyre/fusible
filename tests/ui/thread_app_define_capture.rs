// SPDX-License-Identifier: MPL-2.0
// SPDX-FileCopyrightText: Copyright 2024 Ian McIntyre

//! Make sure that a thread cannot capture the AppDefine reference.
//!
//! If a thread were able to capture the AppDefine reference, it would
//! be able to signal to callees that we're executing in the application
//! definition callback. That wouldn't be true; we'd be executing in
//! a thread!
//!
//! With a little bit of unsafe code, we make sure that this is true
//! no matter the Send and Sync bounds on AppDefine.

use core::pin::pin;
use fusible::thread::{StaticStack, Thread};

static STACK: StaticStack<512> = StaticStack::new();

#[derive(Clone, Copy)]
struct UnsafeAppDefine<'ad, 'pke>(&'ad fusible::AppDefine<'ad, 'pke>);
unsafe impl Sync for UnsafeAppDefine<'_, '_> {}
unsafe impl Send for UnsafeAppDefine<'_, '_> {}
impl<'ad, 'pke> core::ops::Deref for UnsafeAppDefine<'ad, 'pke> {
    type Target = &'ad fusible::AppDefine<'ad, 'pke>;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

fn main() {
    let thread = pin!(Thread::context());
    fusible::kernel_enter(|app_define| {
        let app_define = UnsafeAppDefine(app_define);
        app_define
            .create_thread(
                thread.into_ref(),
                STACK.take().unwrap(),
                &Default::default(),
                move || {
                    let _app_define = app_define;
                },
            )
            .unwrap();
        let _app_define = app_define;
    });
}
