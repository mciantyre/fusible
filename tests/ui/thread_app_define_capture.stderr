error[E0521]: borrowed data escapes outside of closure
  --> tests/ui/thread_app_define_capture.rs:34:9
   |
32 |       fusible::kernel_enter(|app_define| {
   |                              ----------
   |                              |
   |                              `app_define` declared here, outside of the closure body
   |                              `app_define` is a reference that is only valid in the closure body
33 |           let app_define = UnsafeAppDefine(app_define);
34 | /         app_define
35 | |             .create_thread(
36 | |                 thread.into_ref(),
37 | |                 STACK.take().unwrap(),
...  |
41 | |                 },
42 | |             )
   | |_____________^ `app_define` escapes the closure body here
   |
   = note: requirement occurs because of the type `UnsafeAppDefine<'_, '_>`, which makes the generic argument `'_` invariant
   = note: the struct `UnsafeAppDefine<'ad, 'pke>` is invariant over the parameter `'ad`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
